gh-108724: Add PyMutex and _PyParkingLot APIs

colesbury opened the pull request:
`PyMutex` is a one byte lock with fast, inlineable lock and unlock functions for the common uncontended case. The design is based on WebKit's `WTF::Lock`.

PyMutex is built using the `_PyParkingLot` APIs, which provides a cross-platform futex-like API (based on WebKit's `WTF::ParkingLot`). This internal API will be used for building other synchronization primitives used to implement PEP 703, such as one-time initialization and events.

This also includes tests and a mini benchmark in `Tools/lockbench/lockbench.py` to compare with the existing `PyThread_type_lock`.

Uncontended acquisition + release:
Linux (x86-64): PyMutex: 11 ns, PyThread_type_lock: 44 ns
macOS (arm64): PyMutex: 13 ns, PyThread_type_lock: 18 ns
Windows (x86-64): PyMutex: 13 ns, PyThread_type_lock: 38 ns

### PR Overview

The primary purpose of this PR is to implement `PyMutex`, but there are a number of support pieces (described below).

* `PyMutex`: A 1-byte lock that doesn't require memory allocation to initialize and is generally faster than the existing `PyThread_type_lock`. The API is internal only for now.
* `_PyParking_Lot`: A futex-like API based on the API of the same name in WebKit. Used to implement `PyMutex`.
* `_PyRawMutex`: A word sized lock used to implement `_PyParking_Lot`
* `PyEvent`: a one time event. This was used a bunch in the "nogil" fork and is useful for testing the `PyMutex` implementation, so I've included it as part of the PR.
* `pycore_llist.h`: Defines common operations on doubly-linked list. Not strictly necessary (could do the list operations manually), but they come up frequently in the "nogil" fork. (Similar to https://man.freebsd.org/cgi/man.cgi?queue)

### Linked issue

<!-- gh-issue-number: gh-108724 -->
* Issue: gh-108724
<!-- /gh-issue-number -->

colesbury committed:
gh-108724: Add PyMutex and _PyParkingLot APIs

`PyMutex` is a one byte lock with fast, inlineable lock and unlock
functions for the common uncontended case. The design is based on
WebKit's `WTF::Lock`.

PyMutex is built using the `_PyParkingLot` APIs, which provides a
cross-platform futex-like API (based on WebKit's `WTF::ParkingLot`).
This internal API will be used for building other synchronization
primitives used to implement PEP 703, such as one-time initialization
and events.

colesbury committed:
Add news entry

colesbury committed:
Revert change to Modules/_testcapi/parts.h

colesbury committed:
Update Tools/c-analyzer/cpython/ignored.tsv

colesbury committed:
Fix _Py_atomic_store_ptr_release on winarm64

colesbury committed:
Support pthread_stubs.h and require threads for lock tests

colesbury committed:
Clean-up Windows includes and ifdefs

colesbury committed:
Fix Raspbian build

colesbury committed:
Fix wasm tests.

The lock tests were being picked up twice: once in test_lock.py and once
in Test_testinternalcapi from test_misc.py. The Test_testinternalcapi
was not skipping tests when the platform doesn't have threads. This
moves the tests to test_misc.py, doesn't include them in
Test_testinternalcapi, and skips them if the platform doesn't support
threads.

colesbury committed:
Fix indentation

ericsnowcurrently commented on Include/Python.h:
```diff
@@ -48,6 +48,7 @@
 #include "pytypedefs.h"
 #include "pybuffer.h"
 #include "pystats.h"
+#include "cpython/pyatomic.h"
```
Most header files in Include/cpython are pulled in via corresponding files in Include.  Clearly there are some cases where we include "cpython/....h" directly, where the hypothetical .h file in Include/ would only include the file from Include/cpython/.  However, I'm not a fan of that, as I don't think files there should have any awareness of `Py_LIMITED_API`.  I'd rather see this line be `#include "pyatomic.h"` and that we follow the broader pattern.

ericsnowcurrently replied:
Also, this line basically makes everything in Include/cpython/pyatomic.h public API.  We mostly aren't doing leading underscores any more, using a "PyUnstable_" prefix instead.  Public API needs docs, etc. too.

Regardless, does this PR need pyatomic.h in the public API?  If not, why is the file in Include/cpython instead of Include/internal?

(Out of curiosity, how is Include/cpython/pyatomic.h different from pycore_atomic.h?  Could they be unified?)

colesbury replied:
I'll adjust the headers as you request. See the issue #108337 and PR #108338 for a discussion of the motivations and naming. The tl;dr is:
* Yes, it can subsume `pycore_atomic.h` (but I will wait a bit before attempting that)
* It covers atomic operations on more data types, which is needed by PEP 703
* Unlike `pycore_atomic.h`, it can be used in public headers, which will be needed very soon by the biased reference counting implementation

ericsnowcurrently commented on Include/cpython/pyatomic.h:
```diff
@@ -83,6 +83,7 @@
 
+#ifndef Py_LIMITED_API
 #ifndef Py_ATOMIC_H
 #define Py_ATOMIC_H
```
As noted above, I'd rather this followed the broader pattern:

```suggestion
#ifndef Py_CPYTHON_ATOMIC_H
#  error "this header file must not be included directly"
#endif
```

and we would have Include/pyatomic.h as follows:

```C
#ifndef Py_ATOMIC_H
#define Py_ATOMIC_H
#ifdef __cplusplus
extern "C" {
#endif

#ifndef Py_LIMITED_API
#  define Py_CPYTHON_ATOMIC_H
#    include "cpython/pyatomic.h"
#  endif
#endif

#ifdef __cplusplus
}
#endif
#endif  /* !Py_ATOMIC_H */
```

colesbury replied:
I've updated `pyatomic.h` to follow this pattern.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+//
+// Typical initialization:
+//   PyMutex m;
+//   memset(&m, 0, sizeof(m));
```
I'm pretty sure we're moving more toward static initializers:

```suggestion
//   PyMutex m = (PyMutex){0};
```

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+// the mutex is unlocked. If the current thread holds the GIL, then the GIL
+// will be released while the thread is parked.
+static inline void
+PyMutex_Lock(PyMutex *m)
```
I was going to recommend using "acquire"/"release" instead of "lock"/"unlock", as the former is quite prevalent in this context (and, to my brain, "lock" evokes the object rather than the action).  However, there are definitely examples of "lock"/"unlock".

Ultimately, it would be nice if the names were consistent everywhere, to avoid confusing readers.  (It would also be nice if the naming were consistent in the industry and academia.)  The closer we can get, the better.  I'd rather this PR help us converge than diverge, but I haven't taken the time to decide which way the PR goes.😄  (My initial gut reaction was that it diverges.)  Feedback on this from a couple of other core devs would be helpful.

colesbury replied:
Lock/unlock is much more common in libraries and programming languages:
* pthread_mutex_t (POSIX)
* std::mutex (C++)
* std::sync::mutex (Rust)
* java.util.concurrent.locks.Lock (Java)
* sync.Mutex (Go)

Windows tends to use "wait"/"release" terminology.

ericsnowcurrently replied:
FWIW, I'm fine with lock/unlock (vs. acquire/release) at this point.

Regardless, I don't see this as a blocker for the PR, since it isn't public API (yet--and the name would get more scrutiny during the PEP process).

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+    _Py_UNLOCKED = 0,
+    _Py_LOCKED = 1,
+    _Py_HAS_PARKED = 2,
+} _PyMutex_State;
```
What's the advantage of using an enum here over macros, given we're using `uint8_t` as the type in `Py_Mutex` and elsewhere, not `_PyMutexState`?

colesbury replied:
The `typedef ... _PyMutexState` doesn't do anything. I'll remove it. We use enums in [some](https://github.com/python/cpython/blob/1ce9ea0453f7dc69dd41684f3bc9310259513de8/Modules/_pickle.c#L40-L43) [places](https://github.com/python/cpython/blob/1ce9ea0453f7dc69dd41684f3bc9310259513de8/Include/internal/pycore_frame.h#L46-L51) and defines in others to define integer constants. Let me know if you have a preference for `define` over `enum` here.

ericsnowcurrently replied:
In this case I'd personally go with a define.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+//   ...
+//   PyMutex_Unlock(&m);
+typedef struct _PyMutex {
+    uint8_t v;
```
Does this need to be byte-aligned when embedded?

ericsnowcurrently replied:
Also, FWIW, I've come to appreciate a pattern we use elsewhere for bit fields in a struct like this.  Here it would look something like this:

```suggestion
    struct {
        unsigned int locked:1;
        unsigned int has_parked:1;
        /* padding to align to 1 byte */
        unsigned int :6;
    } v;
```

Would that be less efficient or compile differently?

Would it be too clunky (e.g. due to casts to `uint8_t`)?  (I suppose we could use a union to expose it as `uint8_t`.  Is that getting too clever?)

colesbury replied:
It doesn't need any extra care regarding alignment. (Minimum alignment in C is 1 byte -- you can't address or align something smaller than that.)

Using a bitfield ends up being too clunky, because we need to use atomic operations.
- We can't do atomic operations on `struct` types or bitfields, just scalar types because we need to wrap MSVC atomic intrinsics as well as C11 style atomics. 
- The suggested definition is 4 bytes (not 1 byte!) on [GCC, Clang, and MSVC](https://gcc.godbolt.org/z/MfzP6xP7d). It's tricky to get the exact size you want with bitfields and in this case it's important that the mutex is only a single byte.

ericsnowcurrently replied:
Thanks for the explanation.  That makes sense.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+static inline void
+PyMutex_Lock(PyMutex *m)
+{
+    uint8_t expected = _Py_UNLOCKED;
```
Would it make sense to make this const?

```suggestion
    const uint8_t expected = _Py_UNLOCKED;
```

What about static?

FWIW, it isn't obvious to me why we have to pass a pointer for `expected`.

colesbury replied:
It can't be const. The compare_exchange operations take a pointer to "expected". If the comparison fails, `expected` is updated with the current value. That's not relevant here, but can be useful in compare_exchange loops.

In the first version of the pyatomic.h PR, I had compare_exchange operations take a value (not a pointer) for expected, because I thought it was easier to understand, but the feedback from reviewers in #108338 was that it was more important to match the C11 API.

ericsnowcurrently replied:
Thanks for clarifying.  That makes sense.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+}
+
+// Checks if the mutex is currently locked.
+static inline int
+PyMutex_IsLocked(PyMutex *m)
```
I noticed that `WTF::Lock` doesn't have this.  What's the value of it?

colesbury replied:
Mostly for asserting that the lock is actually held, such as `assert(PyMutex_IsLocked(&mutex));` and also will be needed to implement some of the semantics of `_thread.lock`, if and when that's changed to use `PyMutex`. https://github.com/python/cpython/blob/1ce9ea0453f7dc69dd41684f3bc9310259513de8/Modules/_threadmodule.c#L242-L243

ericsnowcurrently replied:
Sound good.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+    _PY_LOCK_DETACH = 1,
+
+    // Handle signals if interrupted while waiting on the lock.
+    _PY_LOCK_MAKE_PENDING_CALLS = 2,
```
Is this only needed for signals or does it extend to all the various things that can interrupt the eval loop (i.e. trigger the eval breaker)?

colesbury replied:
It's only for signals and pending calls (`_PyEval_MakePendingCalls`). The motivation is to have the exact same semantics as `_thread.lock` so that `PyMutex` can be used to implement it. I don't expect this flag to be used outside of `_threadmodule.c`.

https://github.com/python/cpython/blob/1ce9ea0453f7dc69dd41684f3bc9310259513de8/Modules/_threadmodule.c#L108-L113

ericsnowcurrently replied:
Got it.  It would be more clear to name this flag something like "_PY_LOCK_HANDLE_SIGNALS".  The function really should be renamed from `_PyEval_MakePendingCalls()` to `_PyEval_HandlePendingSignals()`.  The relationship with pending calls used to be quite close but has been quite weak for a while.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+    _Py_LOCK_DONT_DETACH = 0,
+
+    // Detach/release the GIL when waiting on the lock.
+    _PY_LOCK_DETACH = 1,
```
I'm not sure why it took me a second to un-misunderstand this, but it did.  Consider something like the following change:

```suggestion
    // Detach/release the GIL while waiting on the lock.
    _PY_LOCK_DETACH = 1,
```

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+// Lock a mutex with an optional timeout and additional options. See
+// _PyLockFlags for details.
+extern PyLockStatus
+_PyMutex_TimedLock(PyMutex *m, _PyTime_t timeout_ns, _PyLockFlags flags);
```
or `_PyMutex_LockTimed()`?  (a little more verb-y)

colesbury replied:
Thanks - I think that's a better name.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+// Wait for the event to be set, or until the timeout expires. If the event is
+// already set, then this returns immediately. Returns 1 if the event was set,
+// and 0 if the timeout expired or thread was interrupted.
+PyAPI_FUNC(int) PyEvent_TimedWait(PyEvent *evt, _PyTime_t timeout_ns);
```
or `PyEvent_WaitTimed()`?

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,161 @@
+// indicate whether the mutex is locked or not. The remaining bits are either
+// zero or a pointer to a `struct raw_mutex_entry` (see lock.c).
+typedef struct {
+    uintptr_t v;
```
Would there be any value in using a union to structurally define the packed representation?

```suggestion
    union {
        uintptr_t addr;
        struct {
            unsigned int locked:1;
        } packed;
    } v;
```

Hmm, maybe that doesn't pay for itself...

colesbury replied:
The union is easier to read, but unfortunately doesn't work here because it's not an "exclusive or". Both the address and locked bit can be valid. Additionally, it's important the locked bit is in the least-significant-bit, but we have not way to guarantee that with bitfields which may be laid out "left to right" or "right to left".

ericsnowcurrently replied:
Yeah, I suspected "endian"-ness might make things messier.

ericsnowcurrently commented on Include/internal/pycore_parking_lot.h:
```diff
@@ -0,0 +1,129 @@
+//          ...
+//      }
+//  });
+#define _PyParkingLot_Unpark(address, unpark, ...)                          \
```
What's the downside to passing a function pointer?  (I'm guessing it's related to how C doesn't have closures.)

colesbury replied:
This could be a function pointer. It may be a little more clunky because of the lack of closures in C, but it's doable. Once you get a chance to look at the usage (in _PyMutex_TryUnlock), let me know if you prefer the function pointer.

ericsnowcurrently commented on Include/internal/pycore_parking_lot.h:
```diff
@@ -0,0 +1,129 @@
+// with respect to unpark operations.
+//
+// The `address_size` argument is the size of the data pointed to by the
+// `address` and `expected` pointers (i.e., sizeof(*address)).
```
FWIW, it's somewhat unclear here what should be passed for "address" and what "address_size" is for.  Not having looked at the implementation yet, I'm guessing there's a `memcmp()` involved.  Regardless, a basic example of usage might be useful here.

colesbury replied:
I'll add an example.

ericsnowcurrently commented on Include/internal/pycore_parking_lot.h:
```diff
@@ -0,0 +1,129 @@
+PyAPI_FUNC(int)
+_PyParkingLot_Park(const void *address, const void *expected,
+                   size_t address_size, _PyTime_t timeout_ns,
+                   void *arg, int detach);
```
There's no description for "detach".  I'm guessing it's something like the `_PY_LOCK_DETACH` flag for `_PyMutex_TimedLock()`.  Is that right?

colesbury replied:
Yes, all the "detach" parameters are basically just to pass the `_PY_LOCK_DETACH` flag down to the code that actually makes the thread wait. I'll add a comment here.

ericsnowcurrently commented on Include/internal/pycore_parking_lot.h:
```diff
@@ -0,0 +1,129 @@
+    // Are there more threads waiting on the address? May be true in cases
+    // where threads are waiting on a different address that maps to the same
+    // internal bucket.
+    int more_waiters;
```
Using a different type (`bool`) or name ("has_more_waiters") might make this a little less ambiguous.  As is, this could easily be read as the number of waiters left or something.

colesbury replied:
I've renamed it to "has_more_waiters"

ericsnowcurrently replied:
Thanks

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+static Bucket buckets[NUM_BUCKETS];
+
+#ifdef HAVE_THREAD_LOCAL
+static _Py_thread_local ThreadData *thread_data = NULL;
```
It would be nice if we didn't have to use a pointer here:

```suggestion
static _Py_thread_local ThreadData thread_data = {0};
```

but we have to in order to init/destroy `thread_data->semas` properly, right?

colesbury replied:
Yes, we would still need to init/destroy `thread_data->semas`. I'm also concerned about compatibility: our other `_Py_thread_local` variables are all pointers and I remember macOS having issues with thread-local aggregate types.

colesbury replied:
I've removed the thread-local data.

ericsnowcurrently commented on Include/internal/pycore_parking_lot.h:
```diff
@@ -0,0 +1,129 @@
+
+// Puts the current thread to sleep until _PySemaphore_Wakeup() is called.
+PyAPI_FUNC(int)
+_PySemaphore_Wait(_PySemaphore *sema, _PyTime_t timeout_ns, int detach);
```
likewise with "detach"

colesbury replied:
I've added a description of `detach`.

ericsnowcurrently reviewed (changes requested):
Here's the first portion of my review, to get you started. 😄

(FWIW, there should probably be a PEP before making a public API out of this, but that doesn't affect this PR.)

colesbury committed:
Rename PyEvent_TimedWait and _PyMutex_LockTimed

colesbury committed:
more_waiters -> has_more_waiters

colesbury committed:
Remove _PyMutex_State typedef

colesbury committed:
Update docs

colesbury committed:
Include cpython/pyatomic.h via pyatomic.h

colesbury committed:
Use compound initializer in test_lock.c

colesbury committed:
Apply suggestions from code review

Co-authored-by: Eric Snow <ericsnowcurrently@gmail.com>

colesbury reviewed (commented):
Thanks @ericsnowcurrently for your feedback. I've replied to the comments so far and updated the PR.

colesbury committed:
Changes from review:

* Rename _PY_LOCK_MAKE_PENDING_CALLS to _PY_LOCK_HANDLE_SIGNALS
* Use define instead of enum for PyMutex states (e.g., _Py_UNLOCKED)

ericsnowcurrently commented on Lib/test/test_capi/test_misc.py:
```diff
@@ -2067,7 +2067,15 @@ def test_version_api_data(self):
+                    and not name.startswith('test_lock_'))
+
+
+@threading_helper.requires_working_threading()
+class Test_PyLock(unittest.TestCase):
+    locals().update((name, getattr(_testinternalcapi, name))
+                    for name in dir(_testinternalcapi)
+                    if name.startswith('test_lock_'))
```
(no action needed)

FYI, I was going to suggest putting the tests in Lib/test/test_lock.py or something.  However, we don't already have distinct tests for `threading.Lock` or for `PyThread_type_lock`, which is a strike against the idea.  Furthermore, this is strictly a C-API feature, so putting the tests in test_capi is probably right. 😄

ericsnowcurrently commented on Lib/test/test_capi/test_misc.py:
```diff
@@ -2067,7 +2067,15 @@ def test_version_api_data(self):
     locals().update((name, getattr(_testinternalcapi, name))
                     for name in dir(_testinternalcapi)
-                    if name.startswith('test_'))
+                    if name.startswith('test_')
+                    and not name.startswith('test_lock_'))
```
(no action needed)

FYI, while looking into this I noticed that we have a bunch of C functions in Modules/_testinternalcapi/pytime.c with the "test_" prefix.  That prefix usually corresponds with a module attribute of the same name, which would get run here as part of this test class.  However, each of the pytime "test_" functions are actually added to the module as "_PyTime_..." attributes.  So this is the first case where we've added C test methods to _testinternalcapi that aren't in Modules/_testinternalcapi.c.  (That isn't a problem, just something I noticed.)

That isn't the case for _testcapimodule.c though.  It pulls in C test methods from a bunch of files in Modules/_testcapi/, which are almost all added to `class Test_testcapi` a little further up here.  That is similar to how you've added C test methods in this PR.  However, in Test_testcapi the `locals().update()` doesn't really exclude any of the tests, as you have, even where those tests are also run elsewhere. [^1]  That's not something to address here though.

Ultimately, I think what you've done is fine, even if not strictly necessary.

[^1]: `grep -ho '{"test_[^"]*' Modules/_testcapi/*.c | awk -F'"' '{print $2}' | xargs -I '{}' grep {} Lib/test/**/*.py`
  Lib/test/datetimetester.py:        _testcapi.test_datetime_capi()
  Lib/test/test_descr.py:                p = property(_testcapi.test_with_docstring)
  Lib/test/test_capi/test_immortal.py:        _testcapi.test_immortal_builtins()
  Lib/test/test_capi/test_immortal.py:        _testcapi.test_immortal_small_ints()
  Lib/test/test_tracemalloc.py:        code = 'import _testcapi; _testcapi.test_pymem_alloc0(); 1'
  Lib/test/test_capi/test_misc.py:        _testcapi.test_widechar()

ericsnowcurrently commented on Lib/test/test_capi/test_misc.py:
```diff
@@ -2067,7 +2067,15 @@ def test_version_api_data(self):
+
+
+@threading_helper.requires_working_threading()
+class Test_PyLock(unittest.TestCase):
```
I'm not sure there's much value in having a distinct Test_PyLock.

Having this distinct test class would definitely make sense if it did more than use the lock-specific tests from _testinternalcapi (like perform some setup/teardown, add additional pure Python tests that called _testinternalcapi helper functions, etc.).  Otherwise Test_testinternalcapi already takes care of everything.  That's the existing pattern: we don't have distinct test classes for other domain-specific test methods.  I'm not clear on the need to do so now for this one case.  Perhaps I've missed something?

colesbury replied:
We skip these tests if there isn't proper threading support, such as on some wasm platforms. (See the `@threading_helper.requires_working_threading()` decorator above.) That's the only reason for pulling these out of `Test_testinternalcapi`.

ericsnowcurrently replied:
Could that be done with `#ifdef WITH_THREAD` in _testinternalcapi.c?

colesbury replied:
No, every build now has `WITH_THREAD` [defined](https://github.com/python/cpython/blob/59073c9ab8eb8db9304f16c46086ccc525e82570/Include/pyport.h#L456-L462) and the check for "Does this platform have working threading?" is in Python.

ericsnowcurrently replied:
yuck

ericsnowcurrently commented on Python/pystate.c:
```diff
@@ -549,6 +550,10 @@ _PyRuntimeState_ReInitThreads(_PyRuntimeState *runtime)
 
     PyMem_SetAllocator(PYMEM_DOMAIN_RAW, &old_alloc);
 
+    // Clears the parking lot. Any waiting threads are dead. This must be
+    // called before releasing any locks that use the parking lot.
+    _PyParkingLot_AfterFork();
```
fork() is so nasty ☹️

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+#define MAX_SEMA_DEPTH 3
+
+typedef struct {
+    Py_ssize_t refcount;
```
We need this because C doesn't offer a hook for thread destruction, right?

colesbury replied:
Yes, but I've now gotten rid of the thread-local data.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    Py_ssize_t refcount;
+
+    int depth;
+    _PySemaphore semas[MAX_SEMA_DEPTH];
```
Comments here would be helpful.

I'm guessing this is strictly the set of pre-initialized semaphores that the current thread may use in the parking lot.

I'm especially interested to understand why more than one is needed, since a thread can only occupy a single slot in the entire parking lot.  (I suppose there are some small races that we must accommodate in the lifecycle of each wait entry relative to what's in the actual buckets.)

Regardless, why 3 specifically?

colesbury replied:
I removed the thread-local data. The idea was to cache _PySemaphore per-thread to avoid initialization of `pthread_mutex_t`/`sem_t`/`Handle`, but after measuring it, the initialization is cheap and much less than the cost of waiting even a brief period (it made no differences in the lock benchmark).

Regarding "3": a thread waiting on a `PyMutex` can end up using two `_PySemaphores`:
1) The parking lot entry uses one
2) If the thread is unluckly, and there's contention on the "buckets" hashtable, the `_PyRawMutex` can also wait on a `_PySemaphore`.

It was 3 instead of 2 because I think the future [critical sections API](https://peps.python.org/pep-0703/#python-critical-sections) can lead to another `_PySemaphore` being used.

ericsnowcurrently replied:
Thanks for the explanation.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+
+    int depth;
+    _PySemaphore semas[MAX_SEMA_DEPTH];
+} ThreadData;
```
(*probably* no action needed)

I was going to say that support for multiple interpreters makes this a bit more interesting.  The current thread may actually have more than one `PyThreadState` associated with it, one for each interpreter running in the thread (only at most one of them is active though). [^1]

However, IIUC, the semaphores here are only in play when a thread has been parked.  If a thread is parked then it is paused and there is no way it could execute code that would switch the thread to another interpreter.  (Thus `ThreadData` can be tied strictly to the thread and doesn't need to relate to each interpreter running in the thread.  Embedding `ThreadData` in `PyThreadState` isn't necessary.)

Otherwise it would require many semaphores and make lifetime management for semaphores messier.  I expect it would also require extra complexity in the parking lot because the design is based on one-entry-per-thread.

[^1]: The current thread state is stored in a thread (`_Py_tss_tstate`).

colesbury replied:
> IIUC, the semaphores here are only in play when a thread has been parked...

Yes (but I've also removed the thread-local data).

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+#define NUM_BUCKETS 251
+
+// Table of waiters (hashed by address)
+static Bucket buckets[NUM_BUCKETS];
```
So we don't ever resize, right?  That implies a very sparse table at first.  It also means there would be increasing numbers of hash collisions as more locks are used. [^1]  That, in turn, means the depth of buckets (and the cost to search for the last lock-matching thread in a bucket) would increase, depending on both the overcommitment of each bucket and on how many threads are sharing a lock.

Without resizing, choosing the right `NUM_BUCKETS` seems pretty important.  Then again, I suppose I don't have any sense of what the actual impact would be of the extra hash collisions.  It would definitely be worth adding a definite statement about that impact to the comment here.

FWIW, I'm not opposed to a static, fixed-size array for the table.  Furthermore, from the `WTL::Lock` blog post, it sounds like resizing is extra messy, so avoiding it isn't necessarily a bad thing. 😄

[^1]: no-gil's one lock per dict, etc. means we'd definitely blow past the threshold...

colesbury replied:
We don't ever resize. As you say, the table resizing is really messy.

The collisions are proportional to the number of **threads** blocked on locks, not the number of locks created. So you'll need thousands of threads before collisions even become a potential problem.

For context, Go has a similar hash table of waiters. They eventually ran into [performance issues](https://github.com/golang/go/issues/17953) that they fixed by using a binary search tree within each bucket. But this was a) 7 years after the original Go release and b) Go can use 100s of thousands of "goroutines", while the practical limits on native threads are much lower.

I don't think this is likely to be a bottleneck anytime soon, but if it turns out to be one, we can do something similar.

ericsnowcurrently replied:
Sounds good.  Thanks for the explanation.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    _PySemaphore semas[MAX_SEMA_DEPTH];
+} ThreadData;
+
+#define NUM_BUCKETS 251
```
Why 251?  It would be worth having a comment here indicating the rationale for the specific number.  IIRC, we have some similar notes in dictobject.c.

colesbury replied:
I'll add a note. The criteria are:

1) Prime number to avoid correlation with common patterns 
2) Aprox. proportional to max number of CPU cores (since each bucket has it's own lock)

The choice of 251 was stolen from Go, which has a similar hash table of waiters:
https://github.com/golang/go/blob/1fde99cd6eff725f5cc13748a43b4aef3de557c8/src/runtime/sema.go#L48-L49. I've changed it to 257 because that makes it easier to statically initialize the buckets.

ericsnowcurrently commented on Include/internal/pycore_lock.h:
```diff
@@ -0,0 +1,158 @@
+// Lightweight locks and other synchronization mechanisms.
+//
+// These implementations are based on WebKit's WTF::Lock. See
+// https://webkit.org/blog/6161/locking-in-webkit/ for a description of the
+// design.
```
Out of curiosity, how closely did you follow the `WTF::Lock` implementation in this PR?  It seems like you diverged in some places.  It may be worth adding comments such places, to note how you've diverged.  That would help any reader that comes in with an understanding of `WTF::Lock`.  Otherwise it's easy, as a reader, to miss the deviations, which can have a negative effect on maintainability down the line.

colesbury replied:
I'm re-reading the blog post and I think all `WTF::Lock` details described in the blog are the same in `PyMutex`.  The `PyMutex_WaitTimed` adds additional functionality that isn't present in `WTF::Lock`, but that's about it.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    if (pthread_cond_init(&sema->cond, NULL)) {
+        Py_FatalError("parking_lot: pthread_cond_init failed");
+    }
+#endif
```
FYI, I don't have much knowledge about these low-level platform-specific APIs (and don't plan on taking the time to become familiar enough to be helpful).  So I'll point out anything I happen to notice, but otherwise I'm trusting that you'll get these parts right or you'll get a review on these parts from someone with more familiarity.  Just a heads up. 😄

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+        }
+        if (err) {
+            res = Py_PARK_TIMEOUT;
+        }
```
What about the signal case.  From thread_pthread.h it looks like it is detectable, so I expect it would be worth reporting.

colesbury replied:
It's handled (res` is initialized to `Py_PARK_INTR` above), but I've made this more explicit.

ericsnowcurrently replied:
perfect!

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    PyThreadState *tstate = _PyThreadState_GET();
+    int was_attached = 0;
+    if (tstate) {
+        was_attached = (tstate->_status.active);
```
We should avoid coupling this to the implementation details of `PyThreadState` (or of the GIL).  Additionally, `tstate->_status.active` does not imply the GIL is held.  Rather, it implies that `_PyThreadState_GET()` would return that tstate, which we already know because it came from there. 😄

It would be better if we used a dedicated function for this:

```suggestion
        was_attached = _PyEval_HoldsLock(tstate);
```

The implementation would go in ceval_gil.c and look something like:

```C
int
_PyEval_HoldsLock(PyThreadState *tstate)
{
    if (((PyThreadState*)_Py_atomic_load_relaxed(&gil->last_holder)) != tstate) {
        return 0;
    }
    return _Py_atomic_load_relaxed(&gil->locked);
}
```

We have a `holds_gil()` in Python/pystate.c, but I'm pretty sure it isn't adequate for our needs here.

colesbury replied:
Ooof, this is tricky. I think the proposed `_PyEval_HoldsLock` has a race condition: 

1) We may be the `last_holder`, but with the GIL unlocked
2) By the time we read `&gil->locked` some other thread may have acquired the GIL so `gil->locked` is 1.

i.e. we may incorrectly think we hold the GIL when some other thread holds it.

I've changed it to just check that there is a non-NULL thread-state. Basically, if `detached` is 1 and there is a non-NULL thread state, then it's now the caller's responsibility to ensure that it holds the GIL.

We'll want to revisit this anyways as further pieces of PEP 703 get integrated, because `--disable-gil` builds will need something like https://peps.python.org/pep-0703/#thread-states.

ericsnowcurrently replied:
> I think the proposed _PyEval_HoldsLock has a race condition:

It works if we flip it around, doesn't it?

```C
int
_PyEval_HoldsLock(PyThreadState *tstate)
{
    if (!_Py_atomic_load_relaxed(&gil->locked)) {
        return 0;
    }
    return tstate == ((PyThreadState*)_Py_atomic_load_relaxed(&gil->last_holder));
}
```

ericsnowcurrently replied:
> if detached is 1 and there is a non-NULL thread state, then it's now the caller's responsibility to ensure that it holds the GIL.

Yeah, that's a pretty common pattern.  Many C-API functions require that the GIL be held.

> We'll want to revisit this anyways as further pieces of PEP 703 get integrated, because --disable-gil builds will need something like https://peps.python.org/pep-0703/#thread-states.

+1

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+}
+
+void
+_PyParkingLot_DeinitThread(void)
```
Likewise, consider "_PyParkingLot_ReleaseThread" [^1] or "_PyParkingLot_MaybeDeinitThread".

```suggestion
_PyParkingLot_ReleaseThread(void)
```

```suggestion
_PyParkingLot_MaybeDeinitThread(void)
```

[^1]: Perhaps that looks too similar to `PyEval_ReleaseThread()`?

colesbury replied:
(removed)

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+}
+
+void
+_PyParkingLot_InitThread(void)
```
Consider "_PyParkingLot_EnsureThread" or "_PyParkingLot_MaybeInitThread", since it doesn't always initialize it.

```suggestion
_PyParkingLot_EnsureThread(void)
```

```suggestion
_PyParkingLot_MaybeInitThread(void)
```

colesbury replied:
Removed along with `thread_data`.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
+        _PySemaphore_Init(&this_thread->semas[i]);
+    }
+    thread_data = this_thread;
```
If it were `static _Py_thread_local ThreadData thread_data = {0};` then we wouldn't need to do so much and could avoid an allocation:

```suggestion
    if (thread_data.refcount == 0) {
        assert(thread_data.depth == 0);
        for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
            _PySemaphore_Init(&thread_data.semas[i]);
        }
    }
    thread_data.refcount++;
```

That said, does that matter enough?  Is there downside to statically allocating the whole `ThreadData` in thread-local memory, rather than just a pointer?

colesbury replied:
(removed)

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+        _PySemaphore_Destroy(&td->semas[i]);
+    }
+
+    PyMem_RawFree(td);
```
Likewise, we could simplify here:

```suggestion
    assert(thread_data.refcount > 0);
    if (--thread_data.refcount == 0) {
        assert(thread_data.depth == 0);
        for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
            _PySemaphore_Destroy(&thread_data.semas[i]);
        }
    }
```

colesbury replied:
(removed)

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    if (!bucket->root.next) {
+        // initialize bucket
+        llist_init(&bucket->root);
+    }
```
FWIW, it would be nice if the linked list didn't leak implementation details like this.  We can be explicit about it:

```suggestion
    if (!llist_is_initialized(&bucket->root)) {
        // initialize bucket
        llist_init(&bucket->root);
    }
```

Otherwise, I'd expect `llist_init()` to do the right thing here (like we do with `_PyParkingLot_InitThread()`:

```suggestion
    // initialize bucket
    llist_init(&bucket->root);
```

or even

```suggestion
    // initialize bucket
    llist_maybe_init(&bucket->root);
```

colesbury replied:
I'd like avoid assigning meaning to `NULL` pointers in the linked-list abstractions, since the meaning is dependent on the context. I've changed the code to statically initialize buckets properly, which avoids the need to lazy-initialize in `enqueue`.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    if (!root->next) {
+        // bucket was not yet initialized
+        return NULL;
+    }
```
likewise:

```suggestion
    struct llist_node *root = &bucket->root;
    if (!llist_is_initialized(root)) {
        return NULL;
    }
```

colesbury replied:
(No longer necessary now that the buckets linked lists are always initialized.)

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+        return NULL;
+    }
+
+    // find the first waiter that is waiting on `address`
```
Notably, it will be the one that's been waiting the longest, right?

colesbury replied:
Yes, at least from the point-of-view of the _PyParkingLot API.

From the point of view of callers to `PyMutex_Lock()`, it might not strictly be the longest waiter because of things like barging: a thread might be woken up but not succeed in acquiring the lock and have to re-queue. That's not too big of a deal because as wait times exceed 1ms we'll start handing off the lock directly.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    case 1: return _Py_atomic_load_uint8(addr) == *(const uint8_t *)expected;
+    case 2: return _Py_atomic_load_uint16(addr) == *(const uint16_t *)expected;
+    case 4: return _Py_atomic_load_uint32(addr) == *(const uint32_t *)expected;
+    case 8: return _Py_atomic_load_uint64(addr) == *(const uint64_t *)expected;
```
Ah, it wasn't clear earlier that we're talking about atomic ints here.  It may be worth making sure it's clear, especially in the header files, that the relevant APIs using `void *` aren't dealing with arbitrary data types.  (It sure would be nice if we could avoid `void *` but I'm guessing we can't.)

colesbury replied:
I don't think it's important that the values are integers. From the API presented by `_PyParkingLot_Park` we just care that `*addr` and `*expected` are bitwise equal (and we only support these specific sizes). They could be pointers (or possibly even float/double). We're using `_Py_atomic_load_uintX` because we need a way to load value at `addr` atomically.

But in practice, all the calls to _PyParkingLot_Park use integers.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+
+// Checks that `*addr == *expected`
+static int
+validate_addr(const void *addr, const void *expected, size_t addr_size)
```
Just to be sure I understood, this actually verifies the values at the addresses match, rather than the addresses themselves.  If that's right, consider changing the name to something more clear, like "validate_at_addrs()".

colesbury replied:
Yes, it's basically `memcmp` but `addr` is loaded atomically. I renamed it to `atomic_memcmp`.

ericsnowcurrently replied:
Yeah, that name is a ton better.  Thanks!

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    if (this_thread->depth >= MAX_SEMA_DEPTH) {
+        Py_FatalError("_PySemaphore_Alloc(): too many calls");
+    }
+    return &this_thread->semas[this_thread->depth++];
```
With a statically allocated `thread_data`:

```suggestion
    if (thread_data.depth >= MAX_SEMA_DEPTH) {
        Py_FatalError("_PySemaphore_Alloc(): too many calls");
    }
    return &thread_data.semas[thread_data.depth++];
```

colesbury replied:
I've removed `thread_data`.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+    this_thread->depth--;
+    if (&this_thread->semas[this_thread->depth] != sema) {
+        Py_FatalError("_PySemaphore_Free(): mismatch wakeup");
+    }
```
With a statically allocated `thread_data`:

```suggestion
    ThreadData *this_thread = thread_data;
    thread_data.depth--;
    if (&thread_data.semas[thread_data.depth] != sema) {
        Py_FatalError("_PySemaphore_Free(): mismatch wakeup");
    }
```

colesbury replied:
I've removed `thread_data`.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+struct wait_entry {
+    struct _PyUnpark unpark;
+    uintptr_t addr;
+    _PySemaphore *sema;
```
What's the advantage of using a pointer here?

colesbury replied:
I've removed the pointer now that the thread-local data is also removed. (It was previously a pointer to one of the cached _PySemaphore in thread-local data.)

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+
+    // timeout or interrupt
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (wait.node.next == NULL) {
```
As with the other cases, consider hiding the implementation detail:

```suggestion
    if (!llist_is_attached(&wait.node)) {
```

colesbury replied:
I've made this a bit more explicit by adding a `is_unparking` field to the wait entry and use that instead of relying on the linked list nodes.

(I don't love `llist_is_attached` since an unatached node in general might have arbitrary `next` and `prev` fields, including possibly, uninitialized memory)

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+void
+_PyParkingLot_FinishUnpark(const void *addr, struct _PyUnpark *unpark)
+{
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
```
If this is always paired with `_PyParkingLot_BeginUnpark()`, we shouldn't need to look up the bucket again:

```C
Bucket *bucket;
struct _PyUnpark *unpark = _PyParkingLot_BeginUnpark(addr, &bucket);
...
_PyParkingLot_FinishUnpark(bucket, unpark);
```

colesbury replied:
I've switched to passing a function pointer to a _PyParkingLot_Unpark call (instead of split begin/end with a macro.). Let me know what you think about the function pointer vs. macro.

ericsnowcurrently replied:
Yeah, that function pointer looks great.

ericsnowcurrently commented on Python/parking_lot.c:
```diff
@@ -0,0 +1,470 @@
+{
+    // After a fork only one thread remains. That thread cannot be blocked
+    // so all entries in the parking lot are for dead threads.
+    memset(buckets, 0, sizeof(buckets));
```
Is there any advantage to this over letting the compiler do it?

```suggestion
    buckets = (Bucket[NUM_BUCKETS]){{{0}}};
```

colesbury replied:
C doesn't let you assign to arrays (and buckets is an array of structs).

It would work if `buckets` were a struct containing an array, but the code generation is [much worse](https://gcc.godbolt.org/z/74oY4T3nz) on MSVC compared to `memset`, even with optimizations enabled. (It uses a *lot* of stack).

ericsnowcurrently reviewed (changes requested):
next batch

colesbury committed:
Remove thread-local data in parking_lot.c

Move _PySemaphore definition to its own header to avoid
leaking platform details in parking_lot.h

colesbury committed:
Add doc for NUM_BUCKETS

colesbury committed:
Make use of Py_PARK_INTR more explicit in _PySemaphore_PlatformWait

colesbury committed:
Statically initialize buckets in parking_lot.c

colesbury committed:
Rename validate_addr to atomic_memcmp

colesbury committed:
Add is_unparking to struct wait_entry in parking_lot.c

colesbury committed:
Use callback in _PyParkingLot_Unpark

colesbury committed:
Simplify _PySemaphore_Wait.

Release the GIL if `detach` is `1` and there is a current thread state.
It's the caller's responsibility to make sure that if there is a valid
thread state (and `detach` is 1), then the current thread holds the GIL.

colesbury reviewed (commented):
Thanks for your continued review. Here's a summary of the major changes:

* I removed the thread-local data from `parking_lot.c`. It wasn't strictly necessary and didn't actually improve performance.
* I switched to statically initializing the `buckets` linked lists in `parking_lot.c`, which avoids needing to check if it's initialized in `enqueue`/`dequeue`.
* I moved the `_PySemaphore` struct definition to `pycore_semaphore.h` because it now includes platform-specific details and I'd like to avoid exposing those in `pycore_parking_lot.h` (I expect more files to use the parking lot API than the `_PySemaphore` API.)
* I switched `_PyParkingLot_Unpark` to use a function pointer as a callback, like you asked about in the previous review. I think this ends up being a bit nicer, but let me know what you think.

colesbury committed:
check-c-globals: increase limit for parking_lot.c

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -0,0 +1,297 @@
+#include <sched.h>          // sched_yield()
+#endif
+
+// If a thread waits on a lock for longer than TIME_TO_BE_FAIR_NS (1 ms), then
+// the unlocking thread directly hands off ownership of the lock. This avoids
+// starvation.
```
Thanks for the nice comment!

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -0,0 +1,297 @@
+// starvation.
+static const _PyTime_t TIME_TO_BE_FAIR_NS = 1000*1000;
+
+// Spin for a bit before parking the thread. This is only enabled for
+// `--disable-gil` builds because it is unlikely to be helpful if the GIL is
+// enabled.
```
another great comment

ericsnowcurrently replied:
FWIW, I'll probably end up using the `PyMutex` API for cross-interpreter operations (e.g. PEP 554).  So even with the GIL this may be helpful.

colesbury replied:
Good to know. We can adjust this then if it's helpful. I expect we'll have some tweaks as we integrate it into various places.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -0,0 +1,297 @@
+    _PyTime_t time_to_be_fair;
+
+    // Set to 1 if the lock was handed off. Written by the unlocking thread.
+    int handoff;
```
```suggestion
    int handedoff;
```

colesbury replied:
Renamed to `handed_off`.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -0,0 +1,297 @@
+#endif
+
+struct mutex_entry {
+    // The time at which the thread should be handed off the lock. Written by
```
FWIW, this reads a bit strange.  It makes more sense if you drop "off".

Also, the wording also implies a specific time, but we're talking about a maximum time after which the thread deserves to get control, right?

colesbury replied:
I've reworded this.

ericsnowcurrently replied:
Thanks!  That's much more clear.

ericsnowcurrently commented on Python/lock.c:
```diff
@@ -0,0 +1,297 @@
+{
+    uint8_t v = _Py_atomic_load_uint8(&m->v);
+    for (;;) {
+        if ((v & _Py_LOCKED) == _Py_UNLOCKED) {
```
FWIW, this is a little confusing at first, as a reader.  We can compare to 0 now that this is a define instead of an enum.

```suggestion
        if ((v & _Py_LOCKED) == 0) {
```

or

```suggestion
        if (!(v & _Py_LOCKED)) {
```

colesbury replied:
I've updated the usages to `if ((v & _Py_LOCKED) == 0) {`

ericsnowcurrently reviewed (changes requested):
Here's the last of my initial review. 😄

Thanks again for doing the work (and for putting up with my many comments)!

colesbury committed:
Changes from review

colesbury reviewed (commented):
Thanks @ericsnowcurrently. I've updated the PR.

ericsnowcurrently reviewed (approved):
LGTM

I'll merge this later today, unless @vstinner, @pitrou, or @gpshead have any objections.
